### PR TITLE
notary: 0.6.1 -> 0.7.0

### PR DESCRIPTION
**Pull request by aaronjheng**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

File: pkgs/tools/security/notary/default.nix
Comment (06kellyjac) [inline pkgs/tools/security/notary/default.nix:53]:
```suggestion
```

Probably best to leave this to `buildGoModule` unless it's narrower than `buildGoModule`

**Review by 06kellyjac — CHANGES_REQUESTED** [review]
In the progress of reviewing the tests

**Comment by 06kellyjac** [comment]
Ok so after my testing I had to skip a godawful amount of tests and some of the errors were heinous such as `panic: crypto/hmac: hash generation function does not produce unique values`. Not something you want in software that's doing a lot of signing.

It seems some of those issues are introduced due to a golang version over 1.15 or so

here's most of the skips and tweaks I had to do :nauseated_face: :

```
  preCheck = ''
    # checkFlags aren't correctly passed through via buildGoModule so we use buildFlagsArray
    # -short only runs the unit-tests tagged short
    buildFlagsArray+="-short"

    substituteInPlace client/backwards_compatibility_test.go \
      --replace "Test0Dot3RepoFormat" "Skip0Dot3RepoFormat"

    substituteInPlace client/client_test.go \
      --replace "TestPublishDelegations(" "SkipPublishDelegations(" \
      --replace "TestPublishDelegationsX509" "SkipPublishDelegationsX509" \
      --replace "TestPublishTargetsDelegationSuccessLocallyHasRoles" "SkipPublishTargetsDelegationSuccessLocallyHasRoles" \
      --replace "TestPublishTargetsDelegationNoTargetsKeyNeeded" "SkipPublishTargetsDelegationNoTargetsKeyNeeded" \
      --replace "TestPublishTargetsDelegationSuccessNeedsToDownloadRoles" "SkipPublishTargetsDelegationSuccessNeedsToDownloadRoles" \
      --replace "TestPublishTargetsDelegationFromTwoRepos" "SkipPublishTargetsDelegationFromTwoRepos" \
      --replace "TestPublishRemoveDelegationKeyFromDelegationRole" "SkipPublishRemoveDelegationKeyFromDelegationRole" \
      --replace "TestPublishRemoveDelegation" "SkipPublishRemoveDelegation" \
      --replace "TestPublishTargetsDelegationCanUseUserKeyWithArbitraryRole" "SkipPublishTargetsDelegationCanUseUserKeyWithArbitraryRole" \
      --replace "TestListRoles" "SkipListRoles"

    # TestSetupCryptoServicesDBStoreSuccess has a panic that looks really bad
    # crypto/hmac: hash generation function does not produce unique values
    substituteInPlace cmd/notary-signer/main_test.go \
      --replace "TestSetupCryptoServicesRethinkDBStoreConnectionFails" "SkipSetupCryptoServicesRethinkDBStoreConnectionFails" \
      --replace "TestSetupCryptoServicesDBStoreSuccess" "SkipSetupCryptoServicesDBStoreSuccess"

    # cirtDir fixes the certificate resolution in the test
    substituteInPlace trustmanager/remoteks/client_test.go \
      --replace "TestErrors" "SkipErrors" \
      --replace "certDir := getCertsDir(t)" "certDir := \"$PWD/fixtures\""

    # All of these have a panic that looks really bad
    # crypto/hmac: hash generation function does not produce unique values
    rm signer/keydbstore/sql_keydbstore_test.go
    # depends on above
    rm \
      server/storage/mysql_test.go \
      signer/keydbstore/mysql_test.go \
      server/storage/postgresql_test.go \
      signer/keydbstore/postgresql_test.go \
      server/storage/sqlite_test.go \
      signer/keydbstore/sqlite_test.go \
      server/storage/sqldb_test.go

    # could not validate the path to a trusted root: unable to retrieve valid leaf certificates
    substituteInPlace trustpinning/certs_test.go \
      --replace "TestValidateRootWithPinnedCA" "SkipValidateRootWithPinnedCA"
  '';
```

---

I tried pulling in the patch that fixes the hash collision but 0.7.0 was too different

I strongly recommend using the latest commit due to this scary error. It also fixes the vast majority of the broken tests

Here's the nix:

(also this is short vs not short, I think not-short is fine)

```
not short - nix-build -A notary  0.41s user 0.09s system 0% cpu 1:40.84 total
    short - nix-build -A notary  0.41s user 0.12s system 0% cpu 1:19.02 total

```

```nix
{ lib, fetchFromGitHub, buildGoModule, fetchpatch, pkcs11Support ? true }:

buildGoModule rec {
  pname = "notary";
  version = "unstable-2022-05-12";

  src = fetchFromGitHub {
    owner = "notaryproject";
    repo = "notary";
    # 0.7.0 is very behind and has some very scary bugs when built with modern go
    # "panic: crypto/hmac: hash generation function does not produce unique values"
    rev = "d7305613db88a58c3343f079906087e44f1d2110";
    sha256 = "sha256-7kUIbm10MXIEKE7MQ+nsptZD8ZAQIK/ZNWK6NIoK3Sw=";
  };
  vendorSha256 = null;

  patches = [
    # fixes Test0Dot3RepoFormat
    (fetchpatch {
      name = "fix-role-parsing-go1.17.patch";
      url = "https://github.com/notaryproject/notary/pull/1635.patch";
      sha256 = "sha256-KHLVCott3185+KqDw/ZqEcbx0e3ObtlJHOz6oRh6hZ0=";
    })
  ];

  excludedPackages = [ "cmd/escrow" ];

  ldflags = [
    "-s"
    "-w"
    "-X github.com/theupdateframework/notary/version.NotaryVersion=${version}"
  ];

  tags = [] ++ lib.optionals pkcs11Support [ "pkcs11key" ];

  preCheck = ''
    # checkFlags aren't correctly passed through via buildGoModule so we use buildFlagsArray
    # -short only runs the unit-tests tagged short
    buildFlagsArray+="-short"
    echo "hi"

    # requires networking
    substituteInPlace cmd/notary-signer/main_test.go \
      --replace "TestSetupCryptoServicesRethinkDBStoreConnectionFails" "SkipSetupCryptoServicesRethinkDBStoreConnectionFails"

    # cirtDir fixes the certificate resolution in the test
    substituteInPlace trustmanager/remoteks/client_test.go \
      --replace "certDir := getCertsDir(t)" "certDir := \"$PWD/fixtures\""
  '';

  meta = with lib; {
    description = "A project that allows anyone to have trust over arbitrary collections of data";
    longDescription = ''
      The Notary project comprises a server and a client for running and
      interacting with trusted collections. See the service architecture
      documentation for more information.

      Notary aims to make the internet more secure by making it easy for people
      to publish and verify content. We often rely on TLS to secure our
      communications with a web server which is inherently flawed, as any
      compromise of the server enables malicious content to be substituted for
      the legitimate content.

      With Notary, publishers can sign their content offline using keys kept
      highly secure. Once the publisher is ready to make the content available,
      they can push their signed trusted collection to a Notary Server.

      Consumers, having acquired the publisher's public key through a secure
      channel, can then communicate with any notary server or (insecure) mirror,
      relying only on the publisher's key to determine the validity and
      integrity of the received content.
    '';
    license = licenses.asl20;
    homepage = "https://github.com/notaryproject/notary";
    maintainers = with maintainers; [ vdemeester ];
  };
}

```

File: pkgs/tools/security/notary/default.nix
Comment (aaronjheng) [reply pkgs/tools/security/notary/default.nix:26]: Nice work. 
Also, I think `buildGoModule` needs some tweaks.

**Comment by aaronjheng** [comment]
Personally, I think we should skip the `check` phase altogether.
1. In this PR, we spent a lot of time in tests which will be a maintainance burden.
2. It makes no sense in packaging.

**Review by 06kellyjac — CHANGES_REQUESTED** [review]
I *prefer* having the checks, it's additional confidence for a tiny build time increase.
Without the tests I wouldn't have noticed that notary is fundamentally broken when built with go 1.15 and higher.
But either way I don't care if we have them disabled.

Notary **must** be updated to this commit or earlier https://github.com/notaryproject/notary/commit/ca095023296d2d710ad9c6dec019397d46bf8576 (19 Oct 2021)

I can not budge on this

```
panic: crypto/hmac: hash generation function does not produce unique values [recovered]
	panic: crypto/hmac: hash generation function does not produce unique values

goroutine 34 [running]:
testing.tRunner.func1.2({0x85c900, 0x9503b0})
	/nix/store/sh0x9kihzkdz15x18ldg989pf29m4nm7-go-1.17.9/share/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/nix/store/sh0x9kihzkdz15x18ldg989pf29m4nm7-go-1.17.9/share/go/src/testing/testing.go:1212 +0x218
panic({0x85c900, 0x9503b0})
	/nix/store/sh0x9kihzkdz15x18ldg989pf29m4nm7-go-1.17.9/share/go/src/runtime/panic.go:1038 +0x215
crypto/hmac.New(0xc0001136c8, {0xc000243cc0, 0xc, 0x10})
	/nix/store/sh0x9kihzkdz15x18ldg989pf29m4nm7-go-1.17.9/share/go/src/crypto/hmac/hmac.go:143 +0x292
github.com/dvsekhvalnov/jose2go/kdf.DerivePBKDF2({0xc000243cc0, 0x3, 0xc000113808}, {0xc000234990, 0x1f, 0x30}, 0x1dfb501606057605, 0x80, {0x962f80, 0xc000026a80})
	/home/jk/go/pkg/mod/github.com/dvsekhvalnov/jose2go@v0.0.0-20170216131308-f21a8cedbbae/kdf/pbkdf2.go:15 +0xb9
github.com/dvsekhvalnov/jose2go.(*Pbse2HmacAesKW).WrapNewKey(0xc00000e0d8, 0xc000108ed0, {0x85c900, 0xc000253190}, 0x1c)
	/home/jk/go/pkg/mod/github.com/dvsekhvalnov/jose2go@v0.0.0-20170216131308-f21a8cedbbae/pbse2_hmac_aeskw.go:50 +0x339
github.com/dvsekhvalnov/jose2go.encrypt({0xc000026a00, 0x79, 0x80}, 0x3, {0x85c900, 0xc000253190})
	/home/jk/go/pkg/mod/github.com/dvsekhvalnov/jose2go@v0.0.0-20170216131308-f21a8cedbbae/jose.go:308 +0x1af
github.com/dvsekhvalnov/jose2go.EncryptBytes({0xc000026a00, 0x79, 0x80}, {0x8d82cf, 0x12}, {0x8d37d3, 0x7}, {0x85c900, 0xc000253190}, {0x0, ...})
	/home/jk/go/pkg/mod/github.com/dvsekhvalnov/jose2go@v0.0.0-20170216131308-f21a8cedbbae/jose.go:225 +0x3a5
github.com/dvsekhvalnov/jose2go.Encrypt(...)
	/home/jk/go/pkg/mod/github.com/dvsekhvalnov/jose2go@v0.0.0-20170216131308-f21a8cedbbae/jose.go:190
github.com/theupdateframework/notary/signer/keydbstore.(*SQLKeyDBStore).AddKey(0xc000166540, {0x8d4ba8, 0x9}, {0x8d27d4, 0x3}, {0x964850, 0xc000207a00})
	/tmp/notary/signer/keydbstore/sql_keydbstore.go:80 +0x155
github.com/theupdateframework/notary/signer/keydbstore.testKeyCanOnlyBeAddedOnce(0x503373, {0x9647b0, 0xc000166540})
	/tmp/notary/signer/keydbstore/keydbstore_test.go:47 +0x12a
github.com/theupdateframework/notary/signer/keydbstore.TestSQLKeyCanOnlyBeAddedOnce(0x0)
	/tmp/notary/signer/keydbstore/sql_keydbstore_test.go:99 +0x74
testing.tRunner(0xc00024eea0, 0x8f37d8)
	/nix/store/sh0x9kihzkdz15x18ldg989pf29m4nm7-go-1.17.9/share/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/nix/store/sh0x9kihzkdz15x18ldg989pf29m4nm7-go-1.17.9/share/go/src/testing/testing.go:1306 +0x35a
```

See also:

https://github.com/dvsekhvalnov/jose2go/issues/26

https://github.com/dvsekhvalnov/jose2go/pull/27

https://github.com/docker/notary-official-images/issues/25

https://github.com/notaryproject/notary/issues/1602

The fix to jose2go is from [1 Sep 2020](https://github.com/dvsekhvalnov/jose2go/pull/27#issue-690050015) which is several months *before* 0.7.0 was tagged ([14 Jan 2021](https://github.com/notaryproject/notary/tags)) but they just didn't update their dependencies

The only way to avoid this panic is to compile the tagged notary with go **1.14** which is not happening, so we need that commit "Update to Go 1.16.9 and bump joes2go" and I tried cherry-picking the commit but that didn't work.

**Comment by aaronjheng** [comment]
For now, I think we should wait for next release.

**Pull request closed without merging**
